launchable/commands/record/build.py

```diff
@@ -169,100 +212,127 @@ def build(ctx: click.core.Context, build_name: str, source: List[str], max_days:
                     name = matched.group('name')
                     if commit_hash and name:
-                        submodules.append((repo_name + "/" + name, repo_dist + "/" + name, commit_hash))
+                        r.append(Workspace(
+                            name=w.name + "/" + name,
+                            dir=w.dir + "/" + name,
+                            commit_hash=commit_hash))
+        return r
+
+    # figure out the commit hash and branch of those workspaces
+    def compute_hash_and_branch(ws: List[Workspace]):
+        # figure out w.commit_hash and w.branch
+        for w in ws:
+            try:
+                if not w.commit_hash:
+                    w.commit_hash = subprocess.check_output("git rev-parse HEAD".split(), cwd=w.dir).decode().replace("\n", "")
+            except Exception as e:
+                click.echo(
+                    click.style(
+                        "Can't get commit hash for {}. Do you run command under git-controlled directory? "
+                        "If not, please set a directory use by --source option.".format(w.dir),
+                        fg='yellow'),
+                    err=True)
+                print(e, file=sys.stderr)
+                sys.exit(1)
+            w.calc_branch_name()
+
+        # Our historical behaviour is to ignore --branch options. In that case, we should reject if that option is present
+        # I also think it's OK to allow this option to explicitly let the branch name set, since the inference
+        #  of the branch name from the workspace is inherently a little fragile
 
-    if len(commits) != 0:
-        invalid = False
-        _commits = normalize_key_value_types(commits)
+    # Rely on --commit and --branch to create a list of workspaces, even when there's no local Git workspaces
+    def synthesize_workspaces() -> List[Workspace]:
+        # workspace by its name
+        ws = {}  # type: Dict[str,Workspace]
 
         commit_pattern = re.compile("[0-9A-Fa-f]{5,40}$")
-        for repo_name, hash in _commits:
+
+        branch_name_map = dict(normalize_key_value_types(branches))
+
+        for name, hash in normalize_key_value_types(commits):
             if not commit_pattern.match(hash):
                 click.echo(click.style(
-                    "{}'s commit hash `{}` is invalid.".format(repo_name, hash),
+                    "{}'s commit hash `{}` is invalid.".format(name, hash),
                     fg="yellow"),
                     err=True)
-                invalid = True
-            submodules.append((repo_name, "", hash))
-        if invalid:
-            sys.exit(1)
+                sys.exit(1)
 
-    # Note: currently becomes unique command args and submodules by the hash.
-    # But they can be conflict between repositories.
-    uniq_submodules = list({commit_hash: (name, repo_dist, commit_hash)
-                            for name, repo_dist, commit_hash, in sources + submodules}.values())
```

Comment:
I also didn't understand this code. I did the code spelunking and traced this all the way back to the original code in 649c1f82e6de9f2f7ce2c32917c5a02690b3439b but it still doesn't make sense.

What it does is that if some of the repositories we have happen to be checking out the same commit hash (so I'm imagining two git submodules pointing to the same commit), then we drop one and only keep one. But that means in one build we can see two sub-modules, the next build we might only see one sub-module, then the next build we might be back to two. That kind of discontinuity would create all sorts of confusion on the server side when we compute CUT.

So I chose not to preserve this behaviour, and instead send all the repositories, even if they point to the same commit. This must be a pretty rare case anyway, so it should be fine.